level00/PropertyTests.hs

```diff
@@ -112,7 +113,45 @@ genTree genKV = fromList <$> Gen.list (Range.linear 0 100) genKV
 genMyBTreeVal :: MonadGen m => m (Int, Char)
 genMyBTreeVal = liftA2 (,) (Gen.int (Range.linear (-100) 100)) (Gen.enum 'a' 'z')
 
--- We're not ready to write a property test for inserting values into our binary
+-- We can also use property based testing to ensure that when we
+-- implement a typeclass instance, that implementation will comply
+-- with the laws of that typeclass. This is important as often the
+-- laws of a typeclass cannot be expressed in the types.
+--
+-- NOTE: There are _two_ valid 'Eq' instances we could write:
```

Comment:
There's also `const $ const True`, and others.

Reply:
Are there other _useful_ instances, you could write one that loops forever, you could leave it as undefined, you could flip the arguments then compare, you could write one that compares the first argument to itself and vice versa then return the XOR of that... 

Happy to expand if there are other *valid* and *useful* instances, but I come down hard on middling "but what if I just X!" pedantry.

Reply:
I don't think it's pedantry. Any function from `a -> a -> Bool` that satisfies the `Eq` laws is valid, as it partitions `a` into equivalence classes. I think comparing keys only gives you a potentially-useful instance, but I haven't thought deeply about that.

I suggest this phrasing:

> The `Eq` instance from `deriving Eq` is too granular: it will consider unequal trees with the same content but different structure, which will be surprising to our callers

Reply:
Write a test that selects the instance that is intended. i.e. all others should fail the test.

Reply:
I've done both. :) Although I suspect the 'you've written the desired instance' test will need expanding as I only counter for structural equality. Should check for key or value only based equality? I don't think participants will be that deliberately tricksy.. ? Maybe?